File: src/main/java/com/telenordigital/prime/events/Storage.java
package com.telenordigital.prime.events;

/**
 * Interface that abstracts the interactions that
 * are necessary to get/update customer data and to both
 * with respect to (slow) accounting, and (fast) provisioning.
 * Typically this interface will represent a fascade towards
 * multiple specialized storage solutions.
 */
public interface Storage  extends ProductDescriptionCache {
    // XXX Shouldn't extend anything I think.

    String injectPurchaseRequest(final PurchaseRequest pr);

    void updatedisplaydatastructure(String msisdn) throws StorageException;

    void removeDisplayDatastructure(String msisdn) throws StorageException;

    void setRemainingByMsisdn(String msisdn, long noOfBytes) throws StorageException;

    Subscriber getSubscriberFromMsisdn(String msisdn) throws StorageException;

    String insertNewSubscriber(String msisdn) throws StorageException;

    void removeSubscriberByMsisdn(String msisdn) throws StorageException;

    void addPurchaseRequestListener(PurchaseRequestListener listener);

    String addRecordOfPurchaseByMsisdn(String ephermeralMsisdn, String sku, long now)
            throws StorageException;

    void removePurchaseRequestById(String id);

    void removeRecordOfPurchaseById(String id);
}

File: prime-api/src/main/kotlin/org/ostelco/prime/storage/legacy/Storage.kt
package org.ostelco.prime.storage.legacy

import org.ostelco.prime.model.ApplicationToken
import org.ostelco.prime.model.Product
import org.ostelco.prime.model.PurchaseRecord
import org.ostelco.prime.model.Subscriber


/**
 * Interface that abstracts the interactions that
 * are necessary to get/update customer data and to both
 * with respect to (slow) accounting, and (fast) provisioning.
 * Typically this interface will represent a facade towards
 * multiple specialized storage solutions.
 */
interface Storage {

    val balances: Map<String, Long>

    /**
     * Get Subscriber Profile
     */
    @Throws(StorageException::class)
    fun getSubscriber(id: String): Subscriber?

    /**
     * Create Subscriber Profile
     */
    @Throws(StorageException::class)
    fun addSubscriber(subscriber: Subscriber): Boolean

    /**
     * Update Subscriber Profile
     */
    @Throws(StorageException::class)
    fun updateSubscriber(subscriber: Subscriber): Boolean

    /**
     * Remove Subscriber for testing
     */
    @Throws(StorageException::class)
    fun removeSubscriber(id: String): Boolean

    /**
     * Link Subscriber to MSISDN
     */
    @Throws(StorageException::class)
    fun addSubscription(id: String, msisdn: String): Boolean

    /**
     * Get Products for a given subscriber
     */
    @Throws(StorageException::class)
    fun getProducts(subscriberId: String): Map<String, Product>

    /**
     * Get Product to perform OCS Topup
     */
    @Throws(StorageException::class)
    fun getProduct(subscriberId: String?, sku: String): Product?

    /**
     * Get balance for Client
     */
    @Throws(StorageException::class)
    fun getBalance(id: String): Long?

    /**
     * Set balance after OCS Topup or Consumption
     */
    @Throws(StorageException::class)
    fun setBalance(msisdn: String, noOfBytes: Long): Boolean

    /**
     * Get msisdn for the given subscription-id
     */
    @Throws(StorageException::class)
    fun getMsisdn(subscriptionId: String): String?

    /**
     * Get all PurchaseRecords
     */
    @Throws(StorageException::class)
    fun getPurchaseRecords(id: String): Collection<PurchaseRecord>

    /**
     * Add PurchaseRecord after Purchase operation
     */
    @Throws(StorageException::class)
    fun addPurchaseRecord(id: String, purchase: PurchaseRecord): String?

    /**
     * Get token used for sending notification to user application
     */
    fun getNotificationTokens(msisdn : String): Collection<ApplicationToken>

    /**
     * Add token used for sending notification to user application
     */
    fun addNotificationToken(msisdn: String, token: ApplicationToken) : Boolean

    /**
     * Get token used for sending notification to user application
     */
    fun getNotificationToken(msisdn: String, applicationID: String): ApplicationToken?

    /**
     * Get token used for sending notification to user application
     */
    fun removeNotificationToken(msisdn: String, applicationID: String): Boolean

    /**
     *
     */
    fun getPaymentId(id: String): String?

    /**
     *
     */
    fun deletePaymentId(id: String): Boolean

    /**
     *
     */
    fun createPaymentId(id: String, paymentId: String): Boolean

    /**
     *
     */
    fun getCustomerId(id: String): String?
}
